Mathlib/CategoryTheory/Shift/InducedShiftSequence.lean

```diff
@@ -0,0 +1,134 @@
+/-
+Copyright (c) 2024 Joël Riou. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Joël Riou
+-/
+import Mathlib.CategoryTheory.Shift.CommShift
+import Mathlib.CategoryTheory.Shift.ShiftSequence
+
+/-! # Induced shift sequences
+
+When `G : C ⥤ A` is a functor from a category equipped with a shift by a
+monoid `M`, we have defined in the file `CategoryTheory.Shift.ShiftSequence`
+a type class `G.ShiftSequence M` which provides functors `G.shift a : C ⥤ A` for all `a : M`,
+isomorphisms `shiftFunctor C n ⋙ G.shift a ≅ G.shift a'` when `n + a = a'`,
+and isomorphisms `G.isoShift a : shiftFunctor C a ⋙ G ≅ G.shift a` for all `a`, all of
+which satisfy good coherence properties. The idea is that it allows to use functors
+`G.shift a` which may have better definitional properties than `shiftFunctor C a ⋙ G`.
+The typical example shall be `[(homologyFunctor C (ComplexShape.up ℤ) 0).ShiftSequence ℤ]`
+for any abelian category `C` (TODO).
+
+Similarly as a shift on a category may induce a shift on a quotient or a localized
+category (see the file `CategoryTheory.Shift.Induced`), this file shows that
+under certain assumptions, there is an induced "shift sequence". The main application
+will be the construction of a shift sequence for the homology functor on the
+homotopy category of cochain complexes (TODO), and also on the derived category (TODO).
+
+-/
+
+open CategoryTheory Category
+
+namespace CategoryTheory
+
+variable {C D A : Type*} [Category C] [Category D] [Category A]
+  {L : C ⥤ D} {F : D ⥤ A} {G : C ⥤ A} (e : L ⋙ F ≅ G) (M : Type*)
+  [AddMonoid M] [HasShift C M] [HasShift D M] [L.CommShift M]
+  [G.ShiftSequence M] (F' : M → D ⥤ A) (e' : ∀ m, L ⋙ F' m ≅ G.shift m)
+  (hL : Nonempty (Full ((whiskeringLeft C D A).obj L)) ∧ Faithful ((whiskeringLeft C D A).obj L))
```

Comment:
Why not `[Full ((whiskeringLeft C D A).obj L)] [Faithful ((whiskeringLeft C D A).obj L)]`?

Reply:
Thanks for the suggestion! Actually, the assumption as `Nonempty Full` instead of `[Full ...]` go back to the time before performance issues led me to make irreducible these definition of the induced shift (and "shift sequences"). As these constructions are used in situations were there is no risk of diamond for the `Full` instance, I have made the change you suggested in both the files `Shift.InducedShiftSequences` and `Shift.Induced`.

Reply:
Hmm, doing so introduced timeouts https://github.com/leanprover-community/mathlib4/actions/runs/8589987780/job/23536915131
Then, I have revert my commit. It seems better if the definitions of induced shift/induced shift sequences do not carry the data of a `Full` instance.

Reply:
Ah ok, thanks for trying!